Implement Kliento site

maribelferreira opened the pull request:
VMKTG-28

- Convert TopNav component to a React component
- Bump versions for @astrojs/mdx, @astrojs/react, and astro in package.json and package-lock.json
- Update Skeleton layout to use TopNav as a React component

maribelferreira committed:
Convert TopNav component to a React component

- Bump versions for @astrojs/mdx, @astrojs/react, and astro in package.json and package-lock.json
- Introduce TopNav component for navigation with mobile support
- Add veraid logo image for branding in the TopNav component
- Update Skeleton layout to use TopNav as a React component

maribelferreira committed:
Update component to include a secondary mobile menu toggle.
- Improved mobile navigation experience with dynamic state management for menu visibility.
- Adjusted styles for better layout and usability.

maribelferreira committed:
Add chevron icon and kliento logo; replace veraid logo in Footer and TopNav

maribelferreira committed:
Refactor navigation components into separate files

- Moved primary and secondary navigation links to constants for better organization.
- Created PrimaryNav and SecondaryNav components to handle respective navigation logic.
- Updated TopNav to utilize new components and manage secondary navigation state.
- Improved code structure and maintainability.

maribelferreira committed:
Refactor TopNav structure and components

- Deleted the old TopNav file and created a new TopNav component to manage primary and secondary navigation.
- Introduced PrimaryNav and SecondaryNav components for better separation of concerns.
- Added constants for navigation links to improve organization.
- Updated Skeleton layout to reflect new TopNav structure.

maribelferreira committed:
Update PrimaryNav styles and clean up code

maribelferreira committed:
Update navigation links and enhance active state styling

- Changed navigation links in PrimaryNav and SecondaryNav to reflect new paths.
- Added current path tracking in PrimaryNav for active link highlighting.
- Improved styling for active links across navigation components.
- Introduced kliento page with relevant content and layout.

maribelferreira committed:
Update secondary navigation links to reflect new paths for kliento

maribelferreira committed:
Remove TopNav index file and clean up PrimaryNav imports

maribelferreira committed:
Enhance navigation components with path tracking and event handling

- Added path tracking in PrimaryNav to set the current path on mount.
- Updated SecondaryNav to set active section based on the current path.
- Refined event handling for service clicks in PrimaryNav.
- Integrated ClientRouter in Skeleton layout for improved navigation transitions.

maribelferreira committed:
Update PrimaryNav styles for active link highlighting

maribelferreira committed:
Refactor navigation link styles in PrimaryNav and SecondaryNav

maribelferreira committed:
Refactor TopNav and PrimaryNav for improved navigation structure

maribelferreira committed:
Enhance PrimaryNav dropdown functionality and styling

maribelferreira committed:
Add navigation to Kliento index page in TopNav component

maribelferreira committed:
Remove unnecessary event prevention in PrimaryNav for Kliento clicks/

maribelferreira committed:
Refactor PrimaryNav and SecondaryNav for service-based navigation

- Updated PrimaryNav to handle service clicks instead of specifically Kliento.
- Enhanced SecondaryNav to filter links based on the current service.
- Introduced dynamic service prefix handling in TopNav for improved navigation structure.

maribelferreira committed:
Refactor navigation components and layouts for improved structure

- Removed TopNav component and integrated PrimaryNav and SecondaryNav into new BaseLayout and ServicePage layouts.
- Updated PrimaryNav to eliminate unnecessary props and handle URL changes dynamically.
- Enhanced SecondaryNav to manage active sections based on the current path.
- Adjusted various pages to utilize the new layouts for better navigation consistency.

maribelferreira committed:
Add navigation hooks and mobile menu toggle for improved responsiveness

- Introduced `useNavigation` hook to manage current path and active link state.
- Created `MobileMenuToggle` component for better mobile navigation experience.
- Refactored `PrimaryNav` and `SecondaryNav` to utilize new hook and mobile toggle.
- Updated styles for active and inactive navigation links in constants file.

maribelferreira committed:
Update Kliento logo image

maribelferreira committed:
Add ServiceNav layout for Kliento page

maribelferreira committed:
Merge branch 'main' into hierarchical-navigation

maribelferreira committed:
Refactor navigation link classes in PrimaryNav and SecondaryNav

- Moved nav link classes from constants file to respective components for better encapsulation.
- Updated class references in PrimaryNav and SecondaryNav to use local definitions.
- Removed unused nav link classes from constants file to streamline code.

maribelferreira committed:
Update font size in Skeleton layout and adjust SecondaryNav for mobile menu toggle

maribelferreira committed:
Clean up navigation hooks in PrimaryNav and SecondaryNav

gnarea commented on src/components/TopNav/SecondaryNav.tsx:
This file is part generic and part Kliento-specific. I think it should be a side, preferably becoming Kliento- and service-agnostic.

gnarea commented on src/components/TopNav/constants.ts:
I don't see the need to split this file and `types.ts`, so maybe combine them into a single `config.ts` file?

gnarea commented on src/components/TopNav/constants.ts:
```diff
@@ -0,0 +1,18 @@
+  { href: "/kliento/clients", text: "Clients", id: "clients" },
+  { href: "/kliento/servers", text: "Servers", id: "servers" },
+  { href: "/kliento/overview", text: "Overview", id: "overview" },
+];
```
You could/should nest the secondary nav items. With the current approach, it'd be impossible to have another section with its own secondary navigation. For example:

```suggestion
export const primaryNavLinks: NavLink[] = [
  { href: "/users", text: "Users" },
  {
    href: "/services",
    text: "Services",
    children: [{
      href: "/kliento",
      text: "Kliento",
      children: [{
        { href: "/clients", text: "Clients", id: "clients" },
        { href: "/servers", text: "Servers", id: "servers" },
        { href: "/overview", text: "Overview", id: "overview" },
        }],
    }],
  },
  { href: "/overview", text: "Tech overview" },
  { href: "/about", text: "About" },
];
```

However, I don't mind this too much because it works, so if you don't want to do it or it's too much work, don't do it.

maribelferreira replied:
I think it makes sense to nest the Kliento secondary navigation items under the Kliento item., which is itself under 'services'. With this approach we'll be able to add another section with its own secondary navigation in the future.

gnarea commented on src/layouts/Base.astro:
Why create this layout instead of updating Skeleton?

gnarea commented on src/layouts/ServicePage.astro:
I think it's an overkill to introduce this layout, which pretty much duplicates `Page.astro` -- it could at least extend it.

I also think it'd make things easier if you make it Kliento-specific:

- Call it `KlientoPage.astro`, extending `Page.astro`.
- Do all the Kliento-specific stuff there, to avoid repetition in the Markdown documents.

gnarea commented on src/pages/kliento/clients.md:
```diff
@@ -0,0 +1,12 @@
+description: "Clients"
+layout: ../../layouts/ServicePage.astro
+service: "Kliento"
+activeSection: "clients"
```
This shouldn't be necessary. We already compute the active page dynamically in the top nav and we could do the same here.

gnarea reviewed (changes requested):
Thanks! UI/UX looks great but I have some suggestions for the implementation.

maribelferreira committed:
Refactor layouts by removing BaseLayout and ServiceNav, introducing KlientoPage for better structure. Updated Page layout to use Skeleton and adjusted index and Kliento pages to reflect new layout changes.

maribelferreira committed:
Fix active section logic in SecondaryNav and remove unused activeSection from Kliento pages for cleaner navigation handling.

maribelferreira committed:
Merge branch 'main' into hierarchical-navigation

maribelferreira committed:
Add navigation configuration file and update imports

- Introduced `config.ts` to define `NavLink` interface and primary/secondary navigation links.
- Updated imports in `PrimaryNav` and `SecondaryNav` to reference the new `config.ts` instead of the removed `constants`.
- Deleted the old `types.ts` file as its content is now integrated into `config.ts`.

maribelferreira committed:
Enhance navigation structure by nesting Kliento service links under the Services menu. Updated `SecondaryNav` to dynamically retrieve these links from `primaryNavLinks`, improving the organization and accessibility of service-related navigation.

maribelferreira committed:
Add scroll behavior to html element in global styles
for consistent vertical scrolling across browsers.

maribelferreira committed:
Refactor `SecondaryNav` to accept `serviceName`, `serviceLinks`, and `logoSrc` as props, enhancing flexibility. Removed the internal link-finding logic in favor of passing links directly. Updated `KlientoPage` to utilize the new props structure, improving navigation organization and maintainability.

maribelferreira committed:
Remove `activeSection` prop from `SecondaryNav` and `KlientoPage` for cleaner code. Updated internal state management in `SecondaryNav` to initialize `localActiveSection` as `null`, simplifying the component's logic.

gnarea commented on src/components/TopNav/SecondaryNav.tsx:
```diff
@@ -1,25 +1,19 @@
 }
 
-export function SecondaryNav({ currentService, activeSection }: SecondaryNavProps) {
+export function SecondaryNav({ serviceName, serviceLinks, logoSrc }: SecondaryNavProps) {
```
This could be further simplified if instead of taking the `serviceName` and `serviceLinks` separately, you take the `serviceNavigation: NavLink`:

```ts
interface SecondaryNavProps {
  nav: NavLink;
  logoSrc: string;
}

export function SecondaryNav({ { href, text, children }: nav, logoSrc }: SecondaryNavProps) {
```

gnarea commented on src/components/TopNav/SecondaryNav.tsx:
```diff
@@ -40,22 +34,18 @@ export function SecondaryNav({ currentService, activeSection }: SecondaryNavProp
-            alt={`${currentService} logo`}
-            className="h-4 md:h-5 w-auto"
-          />
+        <a href={`/${serviceName.toLowerCase()}`}>
```
This is too brittle. That's part of the reason why I suggest to take the whole `NavItem` above, so we won't have to do this.

gnarea commented on src/components/TopNav/SecondaryNav.tsx:
```diff
@@ -28,7 +22,7 @@ export function SecondaryNav({ currentService, activeSection }: SecondaryNavProp
+      const matchedLink = serviceLinks.find((link: NavLink) => path === link.href);
       if (matchedLink && matchedLink.id) {
         setLocalActiveSection(matchedLink.id);
       }
```
I reckon we don't need the `id` attribute because the nav item should be able to figure out whether it's active. Note this is already done in the top nav.

gnarea commented on src/layouts/KlientoPage.astro:
```diff
@@ -0,0 +1,70 @@
+
+interface Props {
+  frontmatter?: Frontmatter;
+  title?: string;
+  description?: string;
+  service?: string;
+  ogImage?: string;
+  ogType?: string;
```
Are these actually used? I reckon everything will work if you deleted these attributes.

gnarea commented on src/layouts/KlientoPage.astro:
```diff
@@ -0,0 +1,70 @@
+  activeSection = null,
+  ogImage,
+  ogType,
+} = schema.parse(Astro.props.frontmatter ?? Astro.props);
```
This layout is only ever used from Markdown, so `Astro.props` is never used.

```suggestion
} = schema.parse(Astro.props.frontmatter);
```

gnarea commented on src/layouts/KlientoPage.astro:
```diff
@@ -0,0 +1,70 @@
+  activeSection: z.string().nullable().optional(),
+  ogImage: z.string().optional(),
+  ogType: z.string().optional(),
+});
```
You don't need to validate things already validated in `Page.astro`.

```suggestion
const schema = z.object({
  service: z.string().optional(),
  activeSection: z.string().nullable().optional(),
});
```

Most importantly, however, `activeSection` is never used and I don't think we need `service` because it's always `Kliento` (that's why we created this template).

gnarea commented on src/layouts/KlientoPage.astro:
```diff
@@ -0,0 +1,70 @@
+interface Frontmatter {
+  title: string;
+  description: string;
+  service?: string;
```
This is always `Kliento` so I think we should drop it.

gnarea commented on src/layouts/KlientoPage.astro:
```diff
@@ -0,0 +1,70 @@
+    client:load
+    serviceName={service}
+    serviceLinks={serviceLinks}
+    logoSrc={service === "Kliento" ? klientoLogo.src : ""}
```
This is always Kliento so let's simplify it.

gnarea commented on src/layouts/KlientoPage.astro:
```diff
@@ -0,0 +1,70 @@
+  return [];
+};
+
+const serviceLinks = findServiceLinks(service);
```
This is the kind of things that `SecondaryNav` should be responsible for. Otherwise, you're going to have to duplicate it in `KlientoPage.astro` and `kliento/index.astro`... As well as any future services.

Also note that doing the lookup by name (`child.text`) is too unreliable. I think we should use `child.href`.

gnarea commented on src/layouts/Page.astro:
```diff
@@ -1,24 +1,32 @@
   title?: string;
   description?: string;
+  ogImage?: string;
+  ogType?: string;
```
I don't think these are used

gnarea commented on src/layouts/Page.astro:
```diff
@@ -1,24 +1,32 @@
 
-const { title, description } = schema.parse(Astro.props.frontmatter ?? Astro.props);
+const { title, description, ogImage, ogType } = schema.parse(
+  Astro.props.frontmatter ?? Astro.props,
```
Same as above:
```suggestion
  Astro.props.frontmatter,
```

gnarea reviewed (changes requested):
Thanks! Left further comments.

Also, note that the active nav item is no longer highlighted.

maribelferreira committed:
Refactor `SecondaryNav` to accept a single `nav` prop, simplifying the component's interface. Updated internal logic to retrieve navigation links from `nav.children`, and modified `KlientoPage` to align with the new structure, enhancing maintainability and clarity.

maribelferreira committed:
Refactor `SecondaryNav` to remove `localActiveSection` and simplify active link handling using `isActive` from `useNavigation`. Clean up `KlientoPage` by removing the `activeSection` property, streamlining the navigation logic and enhancing code clarity.

maribelferreira committed:
Enhance `SecondaryNav` to support dynamic service path handling by integrating `findServiceByPath`. Updated `KlientoPage` to pass `servicePath` prop, streamlining navigation retrieval. Cleaned up `Page` layout by removing unused properties, improving overall code clarity and maintainability.

maribelferreira committed:
Remove unused `service` property from Kliento pages (`clients.md`, `overview.md`, `servers.md`) to enhance code clarity and maintainability.

maribelferreira committed:
Update `KlientoPage`, `Page`, and `Skeleton` to introduce `showBrandInTitle` prop. This allows for conditional rendering of the brand in the title, enhancing flexibility in page presentation.

maribelferreira committed:
Refactor `Kliento` page to use `Skeleton` layout and `SecondaryNav` for the Kliento homepage.

maribelferreira committed:
Fix schema parsing in `Page.astro` to ensure `frontmatter` is correctly utilized, enhancing data validation and consistency.

gnarea commented on src/components/TopNav/SecondaryNav.tsx:
```diff
@@ -1,60 +1,43 @@
 interface SecondaryNavProps {
-  serviceName: string;
-  serviceLinks: NavLink[];
+  nav?: NavLink;
+  servicePath?: string;
```
As discussed, please drop `nav` and make `servicePath` required.

gnarea commented on src/components/TopNav/hooks.ts:
```diff
@@ -27,3 +28,18 @@ export function useNavigation() {
 
   return { currentPath, isActive };
 }
+export function findServiceByPath(servicePath: string): NavLink | null {
```
It's weird to have this function in this file. Did you mean to put it in `config.ts`?

maribelferreira replied:
Yeah, my bad. It should defo be in the config.ts file.

gnarea commented on src/components/TopNav/hooks.ts:
```diff
@@ -27,3 +28,18 @@ export function useNavigation() {
 
   return { currentPath, isActive };
 }
+export function findServiceByPath(servicePath: string): NavLink | null {
```
I think this should throw an error when the nav isn't found.

```suggestion
export function findServiceByPath(servicePath: string): NavLink {
```

gnarea commented on src/layouts/Page.astro:
```diff
@@ -5,28 +5,22 @@ import { z } from "astro:content";
 
 interface Props {
   frontmatter?: Frontmatter;
   title?: string;
   description?: string;
```
Not used

maribelferreira committed:
Refactor `findServiceByPath` to improve error handling and move it from `hooks.ts` to `config.ts` for better organization. Update `SecondaryNav` to directly utilize the new function, ensuring cleaner navigation logic. Remove unused properties from `Page.astro` to enhance clarity and maintainability.

gnarea commented on src/components/TopNav/config.ts:
```diff
@@ -0,0 +1,50 @@
+  { href: "/about", text: "About" },
+];
+
+export function findServiceByPath(servicePath: string): NavLink {
+  const normalizedPath = servicePath.startsWith("/") ? servicePath : `/${servicePath}`;
+
+  const servicesItem = primaryNavLinks.find((item) => item.href === "/services");
+
+  if (!servicesItem) {
+    throw new Error("Services navigation item not found in primary navigation");
+  }
+
+  if (!servicesItem.children || servicesItem.children.length === 0) {
+    throw new Error("No services defined in navigation config");
+  }
+
+  const service = servicesItem.children.find(
+    (child) => child.href.toLowerCase() === normalizedPath.toLowerCase(),
+  );
+
+  if (!service) {
+    throw new Error(`Service not found for path: ${servicePath}`);
+  }
+
+  return service;
+}
```
I think this can be simplified by iterating a single time, and avoiding the path normalisation (I think it's unnecessary; `KlientoPage.astro` and `kliento/index.astro` must just pass a valid path).

```suggestion
export function findServiceByPath(servicePath: string): NavLink {
  const service = primaryNavLinks
    .flatMap((item) => (item.children ? [item, ...item.children] : [item]))
    .find((item) => item.href.toLowerCase() === servicePath.toLowerCase());

  if (!service) {
    throw new Error(`Service not found for path: ${servicePath}`);
  }

  return service;
}
```

(I haven't tested this code)

maribelferreira committed:
Update `findServiceByPath` to simplify error handling and ensure proper service path usage. Adjust `servicePath` in `SecondaryNav` components across `KlientoPage` and `kliento/index.astro` to include leading slashes for consistency and improved navigation functionality.

maribelferreira committed:
Refactor `findServiceByPath` to `findNavItemByPath` for improved navigation item retrieval. Simplified logic to flatten navigation structure and updated error message for clarity. Adjusted `SecondaryNav` to use the new function, enhancing overall navigation handling.

gnarea reviewed (approved):
LGTM! However, I won't merge this PR until we complete the Kliento section. I'll also change the title of this PR to reflect that.

maribelferreira committed:
Hide dropdown menu

gnarea commented on src/components/TopNav/PrimaryNav.tsx:
```diff
@@ -53,12 +57,12 @@ export function PrimaryNav() {
           {primaryNavLinks.map((link) => (
             <li key={link.text} className="relative flex items-center">
-              {link.children ? (
+              {link.children && ENABLE_DROPDOWNS ? (
```
This should suffice to hide the dropdown in child elements. Why are you using `ENABLE_DROPDOWNS` in child elements?

maribelferreira replied:
That's true. I've removed the redundant checks.

gnarea replied:
Although it works, you made the opposite change, which is more complicated.

Rather than removing the parent check, you should've removed the child checks, which is what I suggested because it's simpler and cleaner; you only check once, not several times.

Anyway, this works so I'm merging this branch and we'll delete this functionality soon anyway.

maribelferreira committed:
Removed the redundant checks
